src/vdaf/mastic.rs

```diff
@@ -0,0 +1,651 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Implementation of Mastic as specified in [[draft-mouris-cfrg-mastic-01]].
+//!
+//! [draft-mouris-cfrg-mastic-01]: https://www.ietf.org/archive/id/draft-mouris-cfrg-mastic-01.html
+
+use crate::{
+    codec::{CodecError, Decode, Encode, ParameterizedDecode},
+    field::FieldElement,
+    flp::{
+        szk::{Szk, SzkProofShare},
+        Type,
+    },
+    vdaf::{
+        poplar1::Poplar1AggregationParam,
+        xof::{Seed, Xof},
+        Aggregatable, Client, OutputShare, Vdaf, VdafError,
+    },
+    vidpf::{
+        Vidpf, VidpfError, VidpfInput, VidpfKey, VidpfPublicShare, VidpfServerId, VidpfWeight,
+    },
+};
+use std::ops::{BitAnd, Not};
+use std::{fmt::Debug, io::Cursor};
+use subtle::{Choice, ConstantTimeEq};
+
+/// The main struct implementing the Mastic VDAF.
+/// Composed of a shared zero knowledge proof system and a verifiable incremental
+/// distributed point function.
+#[derive(Clone, Debug)]
+pub struct Mastic<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    algorithm_id: u32,
+    szk: Szk<T, P, SEED_SIZE>,
+    pub(crate) vidpf: Vidpf<VidpfWeight<T::Field>, 16>,
+    /// The length of the private attribute associated with any input.
+    pub(crate) bits: usize,
+}
+
+impl<T, P, const SEED_SIZE: usize> Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Creates a new instance of Mastic, with a specific attribute length and weight type.
+    pub fn new(
+        algorithm_id: u32,
+        szk: Szk<T, P, SEED_SIZE>,
+        vidpf: Vidpf<VidpfWeight<T::Field>, 16>,
+        bits: usize,
+    ) -> Self {
+        Self {
+            algorithm_id,
+            szk,
+            vidpf,
+            bits,
+        }
+    }
+}
+
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation and a set of prefixes to evaluate at that level.
+#[derive(Clone, Debug)]
+pub struct MasticAggregationParam {
+    /// aggregation parameter inherited from [`Poplar1`]: contains the level (attribute length) and a vector of attributes (IdpfInputs)
+    poplar_param: Poplar1AggregationParam,
+    /// Flag indicating whether the VIDPF weight needs to be validated using SZK.
+    root_check_flag: bool,
```

Comment:
For consistency with "require joint rand" elsewhere
```suggestion
    require_check_flag: bool,
```